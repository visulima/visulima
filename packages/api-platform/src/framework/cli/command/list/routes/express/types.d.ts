// eslint-disable-next-line import/no-extraneous-dependencies
import type { IRoute, PathParams } from "@types/express-serve-static-core";
// eslint-disable-next-line import/no-extraneous-dependencies
import { Router } from "express";

export interface Route extends IRoute {
    stack: Layer[];
    metadata?: any;
    name: string;
}

export interface Layer {
    handle?: Route | Router;
    stack: Layer[];
    route: Route;
    name: string;
    params?: PathParams;
    path?: string;
    keys: Key[];
    regexp: ExpressRegex;
    method: string;
}

export interface ExpressRegex extends RegExp {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    fast_slash: boolean;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    fast_star: boolean;
}

export interface RouteMetaData {
    path: string;
    pathParams: Parameter[];
    method: string;
    metadata?: any;
}

export interface Parameter {
    in: string;
    name: string;
    required: boolean;
    [key: string]: any;
}

export interface Key {
    name: string;
    optional: boolean;
    offset: number;
}
