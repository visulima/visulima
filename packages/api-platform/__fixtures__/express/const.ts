import type { ExpressRegex } from "../../src/framework/cli/command/list/routes/express/types";

export const staticPath = /^\/sub-route2\/?(?=\/|$)/i as ExpressRegex;

export const oneDynamicPath = () => {
    return {
        regex: /^\/sub-route\/(?:([^\/]+?))\/?(?=\/|$)/i as ExpressRegex,
        keys: [
            {
                name: "test1",
                optional: false,
                offset: 12,
            },
        ],
    };
};

export const twoDynamicPaths = () => {
    return {
        regex: /^\/sub-sub-route\/(?:([^\/]+?))\/(?:([^\/]+?))\/?(?=\/|$)/i as ExpressRegex,
        keys: [
            {
                name: "test2",
                optional: false,
                offset: 16,
            },
            {
                name: "test3",
                optional: false,
                offset: 31,
            },
        ],
    };
};

export const operationObject = {
    description: "Returns pets based on ID",
    summary: "Find pets by ID",
    operationId: "getPetsById",
    responses: {
        "200": {
            description: "pet response",
            content: {
                "*/*": {
                    schema: {
                        type: "array",
                        items: {
                            $ref: "#/components/schemas/Pet",
                        },
                    },
                },
            },
        },
        default: {
            description: "error payload",
            content: {
                "text/html": {
                    schema: {
                        $ref: "#/components/schemas/ErrorModel",
                    },
                },
            },
        },
    },
};
