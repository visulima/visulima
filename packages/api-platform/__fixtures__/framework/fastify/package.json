{
    "dependencies": {
        "fastify": "^4.17.1"
    }
}
