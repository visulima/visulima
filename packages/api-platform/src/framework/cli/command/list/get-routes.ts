// eslint-disable-next-line import/no-extraneous-dependencies
import { Server } from "@hapi/hapi";
// eslint-disable-next-line import/no-extraneous-dependencies
import type { Express } from "express";
// eslint-disable-next-line import/no-extraneous-dependencies
import type { FastifyInstance } from "fastify";
// eslint-disable-next-line import/no-extraneous-dependencies
import Koa from "koa";

import expressRoutes from "./routes/express-routes";
import fastifyRoutes from "./routes/fastify-routes";
import hapiRoutes from "./routes/hapi-routes";
import koaRoutes from "./routes/koa-routes";
import apiRouteFileParser from "./routes/next/api-route-file-parser";
import collectApiRouteFiles from "./routes/next/collect-api-route-files";
import type { Route } from "./routes/types.d";

export type FrameworkName = "express" | "koa" | "hapi" | "fastify" | "next";
export async function getRoutes(
    appOrPath: string | Express | Koa | Server | FastifyInstance,
    frameworkName: FrameworkName,
    verbose: boolean,
): Promise<Route[] | null> {
    if (frameworkName === "express") {
        return expressRoutes(appOrPath as Express);
    }

    if (frameworkName === "koa") {
        return koaRoutes(appOrPath as Koa);
    }

    if (frameworkName === "hapi") {
        return hapiRoutes(appOrPath as Server);
    }

    if (frameworkName === "fastify") {
        return fastifyRoutes(appOrPath as FastifyInstance);
    }

    if (frameworkName === "next") {
        const apiRouteFiles = await collectApiRouteFiles(appOrPath as string, verbose);

        if (apiRouteFiles.length === 0) {
            throw new Error(`No API routes found, in "${appOrPath}".`);
        }

        return apiRouteFiles.flatMap((apiRouteFile) => apiRouteFileParser(apiRouteFile, appOrPath as string, verbose));
    }

    return null;
}
