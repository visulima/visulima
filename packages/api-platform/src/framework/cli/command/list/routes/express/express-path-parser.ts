import type { Express, Router } from "express";

import pathRegexParser from "./path-regex-parser";
import type { Key, Layer, Parameter, Route, RouteMetaData } from "./types.d";

/**
 * Parses a route object. Route objects are the leafs of an express router tree
 *
 * @param layer The layer of this route object - represents the stack of middleware and other metadata
 * @param keys The full set of keys for this particular route
 * @param basePath The base path as it was initial declared for this route
 * @returns A ExpressPath object holding the metadata for a given route
 */
const parseRouteLayer = (layer: Required<Layer>, keys: Key[], basePath: string): RouteMetaData => {
    const lastRequestHandler = layer.route.stack.at(-1) as Layer;
    const pathParameters: Parameter[] = keys.map((key) => {
        return { in: "path", name: key.name, required: !key.optional };
    });

    const filtered = layer.route.stack.filter((element) => (element.handle as Route).metadata);

    if (filtered.length > 1) {
        throw new Error("Only one metadata middleware is allowed per route");
    }

    const path = (basePath + layer.route.path).replaceAll(/\/{2,}/gu, "/");

    if (filtered.length === 0) {
        return { method: lastRequestHandler.method, path, pathParams: pathParameters };
    }

    return {
        metadata: ((filtered[0] as Layer).handle as Route).metadata,
        method: lastRequestHandler.method,
        path,
        pathParams: pathParameters,
    };
};

/**
 * Recursive traversal method for the express router and middleware tree.
 *
 * @param routes The array of routes to add to
 * @param path The current path segment that we have traversed so far
 * @param layer The current 'layer' of the router tree
 * @param keys The keys for the parameter's in the current path branch of the traversal
 * @returns void - base case saves result to internal object
 */
const traverse = (routes: RouteMetaData[], path: string, layer: Layer, keys: Key[]): void => {
    // eslint-disable-next-line no-param-reassign
    keys = [...keys, ...layer.keys];

    if (layer.name === "router" && layer.handle) {
        layer.handle.stack.forEach((l: Layer) => {
            // eslint-disable-next-line no-param-reassign
            path = path || "";

            traverse(routes, `${path}/${pathRegexParser(layer.regexp, layer.keys)}`, l, keys);
        });
    }

    if (!layer.route || layer.route.stack.length === 0) {
        return;
    }

    routes.push(parseRouteLayer(layer as Required<Layer>, keys, path));
};

// @TODO use this to parse the express swagger

/**
 * Parses an Express app and generates list of routes with metadata.
 *
 * Can Parse:
 *  - Nested Routers and Complex Express Projects
 *  - Optional parameters e.g. /:name?
 *  - Complex Matching routes e.g. /ma*tch, /ex(ab)?mple
 *  - Regex routes e.g. /\/abc|\/xyz/
 *  - Array of paths e.g. app.get(['/abc', '/xyz']) -> /abc,xyz/
 *
 * @param app The Express app reference. Must be used after all routes have been attached
 *
 * @returns List of routes for this express app with meta-data that has been picked up
 */
const expressPathParser = (app: Express): RouteMetaData[] => {
    // eslint-disable-next-line no-underscore-dangle
    const router: Router = app._router || app.router;
    const routes: RouteMetaData[] = [];

    router.stack.forEach((layer: Layer) => {
        traverse(routes, "", layer, []);
    });

    return routes;
};

export default expressPathParser;
