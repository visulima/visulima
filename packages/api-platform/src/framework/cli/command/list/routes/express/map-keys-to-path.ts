import type { ExpressRegex, Key } from "./types.d";

/**
 * Map's the keys/path variables to the regex inside a given path
 *
 * @param layerRegexPath The regex for a router with path parameters
 * @param keys The keys that represent the path parameters
 *
 * @returns The regex for a path variable converted to original string on the express route
 */
const mapKeysToPath = (layerRegexPath: ExpressRegex, keys: Key[]): string => {
    if (keys.length === 0) {
        throw new Error("must include at least one key to map");
    }

    let convertedSubPath = layerRegexPath.toString();

    keys.forEach((key) => {
        convertedSubPath = key.optional
            ? convertedSubPath.replace("(?:\\/([^\\/]+?))?\\", `/:${key.name}?`)
            : convertedSubPath.replace("(?:([^\\/]+?))", `:${key.name}`);
    });

    return convertedSubPath
        .replace("/?(?=\\/|$)/i", "")
        .replace("/^", "")
        .replaceAll("\\", "")
        .replaceAll(/\/{2,}/gu, "/");
};

export default mapKeysToPath;
