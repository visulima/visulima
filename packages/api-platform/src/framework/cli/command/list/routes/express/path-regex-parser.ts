import mapKeysToPath from "./map-keys-to-path";
import type { ExpressRegex, Key } from "./types.d";

/** Parses an express layer's regex and converts it to the original format seen in code.
 *
 * @param layerRegexPath The layer's regex pattern
 * @param keys The keys that represent the layer's path parameters
 * @returns The path string for that layer
 * Code inspired and modify from:
 * https://github.com/expressjs/express/issues/3308#issuecomment-300957572
 */
const pathRegexParser = (layerRegexPath: ExpressRegex | string, keys: Key[]): string => {
    if (typeof layerRegexPath === "string") {
        return layerRegexPath;
    }

    if (layerRegexPath.fast_slash) {
        return "";
    }

    if (layerRegexPath.fast_star) {
        return "*";
    }

    let mappedPath = "";

    if (keys.length > 0) {
        mappedPath = mapKeysToPath(layerRegexPath, keys);
    }

    const match = /^\/\^((?:\\[$()*+./?[\\\]^{|}]|[^$()*+./?[\\\]^{|}])*)\$\//.exec(layerRegexPath
        .toString()
        .replace("\\/?", "")
        .replace("(?=\\/|$)", "$")) as string[];

    if (Array.isArray(match) && match.length > 1) {
        return (match[1] as string).replace(/\\(.)/g, "$1").slice(1);
    }

    if (mappedPath) {
        return mappedPath.slice(1);
    }

    return layerRegexPath.toString();
};

export default pathRegexParser;
