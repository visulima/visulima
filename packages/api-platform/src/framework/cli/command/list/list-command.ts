import { execSync } from "node:child_process";
import { existsSync, rmSync, statSync } from "node:fs";
import { extname, join } from "node:path";
import process from "node:process";

// eslint-disable-next-line import/no-extraneous-dependencies
import chalk from "chalk";

import { getRoutes } from "./get-routes";
import routesGroupBy from "./routes/routes-group-by";
import routesRender from "./routes/routes-render";
import type { Route } from "./routes/types.d";
import { ALLOWED_EXTENSIONS, getApp, getAppWorkingDirectoryPath, getFrameworkName } from "./utils";

interface RenderOptions {
    excludePaths: string[];
    group: string;
    includePaths: string[];
    methods: string[];
}

const listCommand = async (
    framework: "express" | "fastify" | "hapi" | "koa" | "next" | undefined,
    path: string,
    options: Partial<
        RenderOptions & {
            verbose: boolean;
        }
    > = {},
    // eslint-disable-next-line sonarjs/cognitive-complexity
): Promise<void> => {
    const frameworkPath = join(process.cwd(), path);

    if (!existsSync(frameworkPath)) {
        throw new Error("No such file, invalid path provided.");
    }

    const appWorkingDirectoryPath = getAppWorkingDirectoryPath(frameworkPath);

    if (!appWorkingDirectoryPath) {
        throw new Error("Please initialize local package.json.");
    }

    if (framework === undefined) {
        const frameworkName = getFrameworkName(appWorkingDirectoryPath);

        if (!frameworkName) {
            throw new Error("Couldn't detect supported back-end framework.");
        }

        // eslint-disable-next-line no-param-reassign
        framework = frameworkName;
    }

    let routes: Route[] | null = null;

    if (framework === "next") {
        routes = await getRoutes(frameworkPath, "next", options.verbose ?? false);
    } else {
        if (!statSync(frameworkPath).isFile()) {
            throw new Error(`${frameworkPath} is directory, but file expected.`);
        }

        if (!ALLOWED_EXTENSIONS.includes(extname(frameworkPath))) {
            throw new Error("Please specify application .ts/.js/.mjs/.cjs file.");
        }

        const environmentFilePath = `${appWorkingDirectoryPath}/.env`;

        if (existsSync(environmentFilePath)) {
            // Loads environment vars in the current process so application
            // that depends on them can be loaded properly below
            const dotEnvironmentFilePath = `${appWorkingDirectoryPath}/node_modules/dotenv/lib/main.js`;
            const dotenv = await import(dotEnvironmentFilePath);

            dotenv.config({ path: environmentFilePath });
        }

        const isTypeScriptApp = extname(frameworkPath) === ".ts";
        const tscPath = join(appWorkingDirectoryPath, "node_modules/.bin/tsc");

        if (isTypeScriptApp && !existsSync(tscPath)) {
            throw new Error(`Please install typescript in ${appWorkingDirectoryPath}`);
        }

        try {
            if (isTypeScriptApp) {
                // || rm -r ./framework-list removes framework-list directory in case tsc fails

                try {
                    execSync(`${tscPath} --outDir framework-list >&2`, { cwd: appWorkingDirectoryPath });
                } catch (error: any) {
                    // eslint-disable-next-line no-console
                    console.log("TSC compilation failed. Please resolve issues in your project.\n");
                    // eslint-disable-next-line no-console
                    console.log(error);

                    rmSync(join(appWorkingDirectoryPath, "framework-list"), { recursive: true });
                }
            }

            const appJsFilePath = isTypeScriptApp
                ? join(appWorkingDirectoryPath, "framework-list", frameworkPath.replace(appWorkingDirectoryPath, "").replace(".ts", ".js"))
                : frameworkPath;

            const { default: defaultExport } = await import(appJsFilePath);

            routes = await getRoutes(
                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                ["AsyncFunction", "Function"].includes(defaultExport.constructor.name as string) ? await defaultExport() : getApp(defaultExport, framework),
                framework,
                options.verbose ?? false,
            );
        } finally {
            if (isTypeScriptApp) {
                rmSync(join(appWorkingDirectoryPath, "framework-list"), { recursive: true });
            }
        }
    }

    if (routes === null) {
        throw new Error(`Framework "${framework}" is not supported.`);
    }

    if (Array.isArray(options.includePaths) && options.includePaths.length > 0) {
        routes = options.includePaths.flatMap((ipath) => (routes as Route[]).filter((route) => route.path.startsWith(ipath)));
    }

    if (Array.isArray(options.excludePaths) && options.excludePaths.length > 0) {
        routes = options.excludePaths.flatMap((epath) => (routes as Route[]).filter((route) => !route.path.startsWith(epath)));
    }

    if (typeof options.group === "string" && options.group !== "") {
        // eslint-disable-next-line no-console
        console.log();

        const groupedMap = routesGroupBy(routes, (route) => {
            if (options.group === "path") {
                return route.path.replace("/pages", "").split("/")[1];
            }

            return route.tags[0] ?? "unsorted";
        });

        let counter = 0;

        groupedMap.forEach((groupedRoutes, key) => {
            if (counter > 0) {
                // eslint-disable-next-line no-console
                console.log();
            }

            const dotsCount = (process.stdout.columns - 16 - key.length) / 2;
            const dots = dotsCount > 0 ? Array.from({ length: dotsCount }).fill(" ").join("") : "";

            // eslint-disable-next-line no-console
            console.log(dots + chalk.bold.underline(key));

            routesRender(groupedRoutes, options).forEach((renderedRoute) => {
                // eslint-disable-next-line no-console
                console.log(renderedRoute);
            });

            counter += 1;
        });
    } else {
        // eslint-disable-next-line no-console
        console.log();

        routesRender(routes, options).forEach((renderedRoute) => {
            // eslint-disable-next-line no-console
            console.log(renderedRoute);
        });
    }

    // eslint-disable-next-line no-console
    console.log(`\n  Listed ${chalk.greenBright(String(routes.length))} HTTP ${routes.length === 1 ? "route" : "routes"}.\n`);
};

export default listCommand;
