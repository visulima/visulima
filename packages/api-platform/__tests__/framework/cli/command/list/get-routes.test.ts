// eslint-disable-next-line import/no-extraneous-dependencies
import type { Server } from "@hapi/hapi";
import Hapi from "@hapi/hapi";
// eslint-disable-next-line import/no-extraneous-dependencies
import Router from "@koa/router";
// eslint-disable-next-line import/no-extraneous-dependencies
import type { Express } from "express";
import express from "express";
// eslint-disable-next-line import/no-extraneous-dependencies
import type { FastifyInstance } from "fastify";
// eslint-disable-next-line import/no-extraneous-dependencies
import fastify from "fastify";
// eslint-disable-next-line import/no-extraneous-dependencies
import Koa from "koa";
import {
    beforeEach, describe, expect, it,
} from "vitest";

import { getRoutes } from "../../../../../src/framework/cli/command/list/get-routes";

describe("getRoutes", () => {
    describe("express", () => {
        let app: Express;

        beforeEach(() => {
            app = express();

            app.get("/", (_request, response) => response.sendStatus(200));
            app.get("/activity", (_request, response) => response.sendStatus(200));
            // eslint-disable-next-line radar/no-duplicate-string
            app.get("/activity/:id", (_request, response) => response.sendStatus(200));

            // eslint-disable-next-line radar/no-duplicate-string
            app.get("/users", (_request, response) => response.sendStatus(200));
            // eslint-disable-next-line radar/no-duplicate-string
            app.get("/users/:id", (_request, response) => response.sendStatus(200));
            app.put("/users/:id", (_request, response) => response.sendStatus(200));
            // eslint-disable-next-line radar/no-duplicate-string
            app.get("/users/following", (_request, response) => response.sendStatus(200));
        });

        it("Express Routes", async () => {
            const actualRoutesMap = await getRoutes(app, "express", false);

            expect(actualRoutesMap).toMatchSnapshot();
        });
    });

    describe("koa", () => {
        let app: Koa;
        let router: Router;

        beforeEach(() => {
            app = new Koa();
            router = new Router();

            router.get("/", () => {});
            router.get("/activity", () => {});
            router.get("/activity/:id", () => {});

            router.get("/users", () => {});
            router.get("/users/:id", () => {});
            router.put("/users/:id", () => {});
            router.get("/users/following", () => {});

            app.use(router.routes());
            // eslint-disable-next-line no-console
            app.use(() => console.log("Non-router middleware"));
        });

        it("Koa Routes", async () => {
            const actualRoutesMap = await getRoutes(app, "koa", false);

            expect(actualRoutesMap).toMatchSnapshot();
        });
    });

    describe("hapi", () => {
        let app: Server;

        beforeEach(() => {
            app = Hapi.server({
                port: 8080,
                host: "localhost",
            });

            app.route({
                method: "GET",
                path: "/",
                handler: () => null,
            });

            app.route({
                method: "GET",
                path: "/activity",
                handler: () => null,
            });

            app.route({
                method: "GET",
                path: "/activity/:id",
                handler: () => null,
            });

            app.route({
                method: "GET",
                path: "/users",
                handler: () => null,
            });

            app.route({
                method: ["GET", "PUT"],
                path: "/users/:id",
                handler: () => null,
            });

            app.route({
                method: "GET",
                path: "/users/following",
                handler: () => null,
            });
        });

        it("Hapi Routes", async () => {
            const actualRoutesMap = await getRoutes(app, "hapi", false);

            expect(actualRoutesMap).toMatchSnapshot();
        });
    });

    describe("fastify", () => {
        let app: FastifyInstance;

        beforeEach(() => {
            app = fastify();

            app.get("/", (_request, response) => response.status(200));
            app.get("/activity", (_request, response) => response.status(200));
            app.get("/activity/:id", (_request, response) => response.status(200));

            app.get("/users", (_request, response) => response.status(200));
            app.get("/users/:id", (_request, response) => response.status(200));
            app.put("/users/:id", (_request, response) => response.status(200));
            app.get("/users/following", (_request, response) => response.status(200));
        });

        it("Fastify Routes", async () => {
            const actualRoutesMap = await getRoutes(app, "fastify", false);

            expect(actualRoutesMap).toMatchSnapshot();
        });
    });
});
