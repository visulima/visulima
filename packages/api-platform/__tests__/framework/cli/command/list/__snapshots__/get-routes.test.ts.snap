// Vitest Snapshot v1

exports[`getRoutes > express > Express Routes 1`] = `
[
  {
    "file": "unknown",
    "method": "GET",
    "path": "/",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "PUT",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/following",
    "tags": [],
  },
]
`;

exports[`getRoutes > fastify > Fastify Routes 1`] = `
[
  {
    "file": "unknown",
    "method": "GET",
    "path": "/",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD",
    "path": "/",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD",
    "path": "/activity",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD",
    "path": "/activity/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD",
    "path": "/users",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "PUT",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/following",
    "tags": [],
  },
]
`;

exports[`getRoutes > hapi > Hapi Routes 1`] = `
[
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/activity/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "GET",
    "path": "/users/following",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "PUT",
    "path": "/users/:id",
    "tags": [],
  },
]
`;

exports[`getRoutes > koa > Koa Routes 1`] = `
[
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/activity",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/activity/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/users",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "PUT",
    "path": "/users/:id",
    "tags": [],
  },
  {
    "file": "unknown",
    "method": "HEAD|GET",
    "path": "/users/following",
    "tags": [],
  },
]
`;
